### PR TITLE
WIP Docker

### PR DESCRIPTION
**Pull request by shughes-uk**
Quick dockerfile for turbot.

Definitely not optimized and creates a 1.28gb image, I suspect a lot of that is the scientific libraries though.

 Needs instructions on use too.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,7 @@
+FROM python:3.7
+RUN pip install poetry
+WORKDIR /turbot
+COPY . ./
+RUN poetry config virtualenvs.create false \
+    && poetry install --no-ansi --no-interaction
```
Comment (lexicalunit) [inline Dockerfile:6]: You probably also want to add `--no-dev` to reduce docker image size. The dev dependencies are only required for running the scripts and test suite.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,7 @@
+FROM python:3.7
+RUN pip install poetry
+WORKDIR /turbot
+COPY . ./
+RUN poetry config virtualenvs.create false \
```
Comment (lexicalunit) [inline Dockerfile:5]: I'm surprised this works. Doesn't this tell poetry not to create a venv if one doesn't exist? So does one exist already somehow I guess?

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,7 @@
+FROM python:3.7
```
Comment (lexicalunit) [inline Dockerfile:1]: Since the project supports 3.7+ do we want to default to the latest version here and use 3.8?

Comment (lexicalunit) [inline Dockerfile:1]: Might try and see if the slim version, eg `python:3.x-slim`, works. If it does that would save a ton of space.

**Comment by lexicalunit** [comment]
WDYT about using `pip install .` instead of installing it off PyPI?

**Comment by lexicalunit** [comment]
Might also be worth reading thru https://pythonspeed.com/articles/pipenv-docker/ and especially the section titled `Fast reproducible Docker builds with poetry`. Maybe to support this we could add `poetry export -f requirements.txt -o requirements.txt` to the `publish.sh` script.

One thing I've been thinking about how to solve for a while now is how to make it so turbot knows if it's on a release or if it's on a development branch (or on master with changes past the last release). Right now if you do a `turbot -v` it'll give you the version of the last published release, even if there's been a ton of changes since then. Ideally I'd want it to be able to somehow know that it's a different version if things have been changed. Whatever process allowed this to be done would need to ensure that the generate `requirements.txt` file was also kept up-to-date so it would have to be done somewhere other than in `publish.sh`. Maybe as a git push hook perhaps? Dunno.

File: Dockerfile (hunk 2 above)
Comment (shughes-uk) [reply Dockerfile:5]: it just uses the "system" python

File: Dockerfile (hunk 3 above)
Comment (shughes-uk) [reply Dockerfile:1]: ideally we would use a multi-stage build, create wheels in the 'full fat' initial stage, and then copy them out + install in the 'slim' stage. I was unable to figure out how to do that with poetry, and trying it with pip introduced an issue with being unable to find the vendorized dep

**Comment by shughes-uk** [comment]
So my first port of call was trying `pip install .` but it does not seem to be able to handle the vendorized dep and breaks. Export could maybe work but I think it'll run into the same issue.

File: Dockerfile (hunk 3 above)
Comment (lexicalunit) [reply Dockerfile:1]: Yeah vendoring sucks. In the latest version I totally removed it so it should work with `pip` now.

File: Dockerfile (hunk 2 above)
Comment (lexicalunit) [reply Dockerfile:5]: Fancy.

**Comment by lexicalunit** [comment]
@shughes-uk Sorry, try again now that I've removed the vendorized dep. No idea why that just simply didn't work.

File: Dockerfile (hunk 3 above)
Comment (shughes-uk) [reply Dockerfile:1]: sweet i'll give it a pull

**Pull request merged**